### PR TITLE
cuHyena FFTConv support for Hyena Long Implicit (LI) Layer

### PR DESCRIPTION
**Pull request by farhadrgh**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py [hunk 1]
```diff
@@ -949,11 +955,18 @@
...
 
-            x1, x2, v = [
-                AllToAllSingleFunction.apply(tensor, cp_group, "split_to_full", True) for tensor in [x1, x2, v]
-            ]
+            if inference_context is not None:  # reconstruct ALL tensors from split to full
+                x1, x2, v = [
+                    AllToAllSingleFunction.apply(tensor, cp_group, "split_to_full", True) for tensor in [x1, x2, v]
+                ]
+            else:  # only reconstruct z (post gating) for non-inference case
+                z = AllToAllSingleFunction.apply(z, cp_group, "split_to_full", True)
```
[the remark this answers is not in the document]
Comment (farhadrgh) [reply nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py:983]: fixed in [fc900ed](https://github.com/NVIDIA/NeMo/pull/14396/commits/fc900ed9a982b41dac53c10ad9f0ad8500217b4b)

File: nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py [hunk 2]
```diff
@@ -507,9 +521,13 @@ def __init__(
...
 
-            R = 1e-1 * torch.randn(d_model, order, dtype=torch.float32) / math.sqrt(order)
+            R = (
+                1e-1
+                * torch.randn(d_model, order, dtype=torch.float32, device=torch.cuda.current_device())
+                / math.sqrt(order)
+            )
             self.R = nn.Parameter(R)
-            self.p = nn.Parameter(-torch.ones(d_model, order, dtype=torch.float32))
+            self.p = nn.Parameter(-torch.ones(d_model, order, dtype=torch.float32, device=torch.cuda.current_device()))
```
Comment (jstjohn) [inline nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py]: I don't love setting the device like this here. This relies on the env having properly set current_device etc. What if you wanted to initialize on meta or something? Please either remove or add a comment about what is being fixed. If there's another tensor in context that you want to do a "same device" placement then I would recommend a ones like or an `other.device`.

File: tests/collections/llm/gpt/model/test_hyena_mixer_cp.py [hunk 3]
```diff
@@ -166,18 +166,18 @@ def zigzag_gather_from_group_ranks(data, group, seq_dim=0):
     return reconstructed_data
 
 
-class B2BConv1d(torch.nn.Module):
-    def __init__(self, seq_len, use_b2b_causal_conv1d=False):
+class MixerModuleWrapper(torch.nn.Module):
```
Comment (jstjohn) [inline tests/collections/llm/gpt/model/test_hyena_mixer_cp.py:169]: Feel free to resolve if hard -- would it be too much extra work to copy this into a `_tp.py` test variant or could you sneak in a similar tp test?

Comment (jstjohn) [reply tests/collections/llm/gpt/model/test_hyena_mixer_cp.py:169]: tp is the other main kind of parallelism our models (esp the very long context) uses, so is important that we have tests like this someday. If you resolve, if you could add a ticket to do this later, that would be awesome.

File: nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py (hunk 2 above)
Comment (farhadrgh) [reply nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py]: yeah, the ImplicitModalFilter and ExplicitSingleDecayFilter filters register a few parameters inside their init and they all don't inherit device from the device from module. This change was just one instance that was missing and causing a device missmatch. Also, there is a comment `# Do not register into buffer, so it doesn't cast to BF16!` that's why I hold off of doing that.

File: tests/collections/llm/gpt/model/test_hyena_mixer_cp.py (hunk 3 above)
Comment (farhadrgh) [reply tests/collections/llm/gpt/model/test_hyena_mixer_cp.py:169]: My plan is to have separate test for TP since its not controlled by a module-level arg (e.g. `_use_cp`) and rather is assigned when initializing the `parallel_state`. So I may need to cache on two distinct parallel state and then compare.

File: nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py [hunk 4]
```diff
@@ -1251,9 +1257,12 @@ def forward(self, x, inference_context=None, _use_cp=True):
         # maybe handle num_groups
         weight = weight.repeat_interleave(self.group_dim, dim=0)
 
-        if self.use_fast_causal_conv:
-            y = causal_conv1d_fn(x, weight, bias=None, activation=None)[..., pad_size:]
-        else:
+        if self.use_fast_causal_conv:  # hyena_proj_conv case
```
Comment (moradza) [inline nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py:1267]: Maybe this one is cleaner.
```
if self.use_cuhyena:  # hyena_proj_conv of LI layer when cuhyena is enabled
       y = causal_conv1d(x, weight)[..., pad_size:]
 elif self.use_fast_causal_conv:  # hyena_proj_conv case
            y = causal_conv1d_fn(x, weight, bias=None, activation=None)[..., pad_size:]
  else:  # hyena_short_conv case
        ```

Comment (farhadrgh) [reply nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py:1267]: I made some changes here: [57eb1eb](https://github.com/NVIDIA/NeMo/pull/14396/commits/57eb1eb50dd8651725bf9fafebd7302d34c4f8a5)

Comment (farhadrgh) [reply nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py:1267]: and here: [a367f75](https://github.com/NVIDIA/NeMo/pull/14396/commits/a367f7517e222a590a04cc6f70b739a61e1a9080)

File: nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py [hunk 5]
```diff
@@ -1227,9 +1260,12 @@ def forward(self, x, inference_context=None, _use_cp=True):
...
+            if self.use_cuhyena:  # hyena_proj_conv of LI layer when cuhyena is enabled
+                y = causal_conv1d(x, weight)[..., pad_size:]
+            else:
+                y = causal_conv1d_fn(x, weight, bias=None, activation=None)[..., pad_size:]
```
Comment (moradza) [inline nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py:1273]: @farhadrgh let's drop padding when we don't have cp.
```

        x_shape = x.shape
        weight = self.short_conv_weight
        pad_size = self.kernel_size - 1
        
        # maybe handle num_groups
        weight = weight.repeat_interleave(self.group_dim, dim=0)
        

        if _use_cp and get_context_parallel_world_size() > 1:

            cp_group = get_context_parallel_group()
            cp_rank = get_context_parallel_rank()

            # Transfer patches across ranks.
            seq_dim = 2  # Last dimension.
            chunk_a, chunk_b = zigzag_get_overlapping_patches(x, seq_dim=seq_dim, overlap_size=pad_size)
            received_a, received_b = ExchangeOverlappingRegionsCausal.apply(chunk_a, chunk_b, cp_group, cp_rank)

            # Pad and rearrange
            x = rearrange(x, "b h (nc s) -> (nc b) h s", nc=2)
            padding = torch.concat([received_a, received_b], dim=0)

            x = torch.concat([padding, x], dim=-1)
            
            if self.use_fast_causal_conv:  # hyena_proj_conv case
                if self.use_cuhyena:  # hyena_proj_conv of LI layer when cuhyena is enabled
                    y = causal_conv1d(x, weight)[..., pad_size:]
                else:
                    y = causal_conv1d_fn(x, weight, bias=None, activation=None)[..., pad_size:]
            else:  # hyena_short_conv case

                y = F.conv1d(
                    x,
                    weight,
                    bias=None,
                    stride=1,
                    padding=0,
                    groups=self.conv_groups,
                )

        else:
            if self.use_fast_causal_conv:  # hyena_proj_conv case
                if self.use_cuhyena:  # hyena_proj_conv of LI layer when cuhyena is enabled
                    y = causal_conv1d(x, weight)
                else:
                    y = causal_conv1d_fn(x, weight, bias=None, activation=None)
            else:  # hyena_short_conv case

                y = F.conv1d(
                    x,
                    weight,
                    bias=None,
                    stride=1,
                    padding=pad_size,
                    groups=self.conv_groups,
                )[..., -pad_size:]

        if _use_cp and get_context_parallel_world_size() > 1:
            y = rearrange(y, "(nc b) h s -> b h (nc s)", nc=2)

        assert y.shape == x_shape, f"y.shape = {y.shape} | x.shape = {x_shape}"

        return y
```

Comment (farhadrgh) [reply nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py:1273]: @moradza Yeah I didn't add this because of repetitive code and convolutted statements. Also this snippet breaks the assert statement in the last line

Comment (farhadrgh) [reply nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py:1273]: Found a better way of doing this here: [d302175](https://github.com/NVIDIA/NeMo/pull/14396/commits/d302175ffd8003db2fd78e00cc64ab4abd8f0a7c)

**Comment by chtruong814** [comment]
Merging this despite the codecov.  codecov is close to target 80% on patch changes and above average.  Challenging to add more tests because the expected dependency is not available yet to be testing further.
